core/runtime/common/module_repository_impl.cpp

```diff
@@ -70,14 +72,18 @@ namespace kagome::runtime {
     // either found existed or emplaced a new one
     BOOST_ASSERT(thread_local_cache != instances_cache_.end());
 
+    if (single_module_cache_ && single_module_cache_->module.has_value()) {
+      BOOST_VERIFY(modules_.put(state, single_module_cache_->module.value()));
+      single_module_cache_->module.reset();
```

Comment:
I think we can make the module private and add an extract() method which would both return the module and reset it in the cache, to highlight its intended usage.

Reply:
Did that